media/src/app.ts

```diff
@@ -0,0 +1,276 @@
+console.log(`Starting with ID ${lastProcessedId}`);
+
+const saveEntry = (
+  mediaType: MediaType,
+  outputPath: string,
+  outputFilename: string,
+  mediaSrc: string,
+  itemType: ItemType,
+  itemId: number
+) => {
+  let dimensions = {width: 0, height: 0};
+  if (mediaType === MediaType.IMG) {
+    try {
+      dimensions = sizeOf(outputPath);
+    } catch (e) {
+    }
+  }
+  let line = [
+    itemType, itemId, mediaType, mediaSrc, outputFilename, dimensions.width, dimensions.height
+  ];
+  fs.appendFileSync(config.dbFile, line.join(',') + '\n');
+  fs.writeFileSync(config.lastProcessedIdFile, itemId.toString());
+};
+
+const sleep = ms => new Promise(r => setTimeout(r, ms));
+
+const hideUrl = (url: string): string => {
+  try {
+    // parse url
+    const parsedUrl = new URL(url);
+
+    // replace second half of the characters in the string with asterisks
+    const hideChars = (str: string) => {
+      if (!str) {
+        return str;
+      }
+      const half = Math.ceil(str.length / 2);
+      return str.substring(0, half) + '*'.repeat(str.length - half);
+    }
+
+    parsedUrl.username = hideChars(parsedUrl.username);
+    parsedUrl.password = hideChars(parsedUrl.password);
+    parsedUrl.hash = hideChars(parsedUrl.hash);
+    if (parsedUrl.pathname) {
+      // preserve the extension if any
+      const ext = parsedUrl.pathname.match(/^(.*)(\.\w+)$/);
+      if (ext) {
+        parsedUrl.pathname = hideChars(ext[1]) + ext[2];
+      } else {
+        parsedUrl.pathname = hideChars(parsedUrl.pathname);
+      }
+    }
+    parsedUrl.search = hideChars(parsedUrl.search);
+    return parsedUrl.toString();
+  } catch (e) {
+    return url;
+  }
+}
+
+const download = async (
+  url: string,
+  dest: string,
+  signal: AbortSignal
+): Promise<boolean> => {
+  if (!url.match(/^(\/\/|http:|https:)/)) {
+    console.log(`Invalid URL: ${url}`);
+    return false;
+  }
+  if (url.match(/^\/\//)) {
+    url = 'https:' + url;
+  }
+  try {
+    const response = await fetch(url, { signal });
+    const responseBinary = await response.buffer();
+    await fs.promises.writeFile(dest, responseBinary);
+    console.log(`Wrote ${responseBinary.length} bytes to ${dest} for ${hideUrl(url)}`);
+    return true;
+  } catch (e) {
+    console.log(e);
+    return false;
+  }
+};
+
+(async () => {
+  const connection = await mysql.createConnection({
+    host: config.mysql.host,
+    user: config.mysql.user,
+    password: config.mysql.password,
+    database: config.mysql.database
+  });
+
+  const handler = async (startWithId: number) => {
+    const [rows] = await connection.query<IContentSourceItem[]>(
+      `select
+        cs.content_source_id,
+        coalesce(p.html, c.html) as html,
+        coalesce(p.source, c.source) as source,
+        cs.ref_type,
+        cs.ref_id
+      from
+          content_source cs
+      left join comments c on c.content_source_id = cs.content_source_id
```

Comment:
A reminder to remove this join. We want to parse all `content_source` entries, even those from the history (that don't have associated post or comment).